torch/optim/sparse_adam.py

```diff
@@ -75,17 +75,22 @@ def make_sparse(values):
                 #      old <- b * old + (1 - b) * new
                 # <==> old += (1 - b) * (new - old)
                 old_exp_avg_values = exp_avg._sparse_mask(grad)._values()
-                exp_avg.add_(make_sparse(grad_values.sub(old_exp_avg_values).mul_(1 - beta1)))
+                exp_avg_update_values = grad_values.sub(old_exp_avg_values).mul_(1 - beta1)
+                exp_avg.add_(make_sparse(exp_avg_update_values))
                 old_exp_avg_sq_values = exp_avg_sq._sparse_mask(grad)._values()
-                exp_avg_sq.add_(make_sparse(grad_values.pow(2).sub_(old_exp_avg_sq_values).mul_(1 - beta2)))
+                exp_avg_sq_update_values = grad_values.pow(2).sub_(old_exp_avg_sq_values).mul_(1 - beta2)
+                exp_avg_sq.add_(make_sparse(exp_avg_sq_update_values))
 
-                numer = exp_avg._sparse_mask(grad)._values()
-                denom = exp_avg_sq._sparse_mask(grad)._values().sqrt_().add_(group['eps'])
+                # Dense addition again is indended, avoiding another _sparse_mask
+                numer = exp_avg_update_values.add_(old_exp_avg_values)
+                denom = exp_avg_sq_update_values.add_(old_exp_avg_sq_values).sqrt_().add_(group['eps'])
 
                 bias_correction1 = 1 - beta1 ** state['step']
                 bias_correction2 = 1 - beta2 ** state['step']
                 step_size = group['lr'] * math.sqrt(bias_correction2) / bias_correction1
 
                 p.data.add_(make_sparse(-step_size * numer.div_(denom)))
 
+                del exp_avg_update_values, exp_avg_sq_update_values
```

Comment:
oh the del was only meant to be close to the place where we overwrite the `update_values`, because they are no longer what the name suggests. It's not needed if you put it at the end

Reply:
Oh gotcha! :)